### PR TITLE
proxy: Add a way to serve pprof data over HTTP

### PR DESCRIPTION
**Pull request by dlespiau**
The -pprof flag will make the proxy server profiling data overt HTTP.
A nice benefit from having this is that we can always ask for the
backtraces of all goroutines:

  http://localhost:6060/debug/pprof/goroutine?debug=2

This is very handy to debug odd things happening to the proxy and check
we don't have a goroutine blocked somewhere.

Signed-off-by: Damien Lespiau <damien.lespiau@intel.com>

### CODE REVIEW

**Comment by chavafg** [comment]
qa-failed

**Comment by dlespiau** [comment]
```
not ok 42 start and kill state
```

**Comment by sameo** [comment]
LGTM

File: proxy/proxy.go [hunk 1]
```diff
@@ -350,11 +352,26 @@ func initLogging() {
...
+
+	go func() {
+		http.ListenAndServe("localhost:6060", nil)
+	}()
+}
+
 func main() {
 	initLogging()
 
+	flag.BoolVar(&enablePprof, "pprof", false, "enable pprof (visit http://localhost:6060/debug/pprof)")
```
Comment (jodh-intel) [inline proxy/proxy.go]: Could we use a variable to store the port number rather than using another hard-coded string?

Comment (jodh-intel) [inline proxy/proxy.go]: Also, what about an option to specify the port number?

Comment (jcvenegas) [inline proxy/proxy.go]: and ip to listen as well, would help to not only listen in localhost if it is a remote server where I want to debug

File: proxy/proxy.go [hunk 2]
```diff
@@ -350,11 +352,26 @@ func initLogging() {
...
 
+var enablePprof bool
+
+func setupPprof() {
+	if !enablePprof {
+		return
+	}
+
+	go func() {
+		http.ListenAndServe("localhost:6060", nil)
```
Comment (sboeuf) [inline proxy/proxy.go]: @jodh-intel , what are you suggesting instead ? bigger port number instead ?

Comment (sboeuf) [inline proxy/proxy.go]: sorry, didn't read below comments...

Comment (jodh-intel) [inline proxy/proxy.go]: I'm saying `6060` isn't in `/etc/services`, so this number seems to have been randomly chosen. So, what if another application running on the same machine also tries to use the same port number? If it's not a well-known port number (or even if it is infact), there should be a way to change it.

Comment (sboeuf) [inline proxy/proxy.go]: Yes agreed, and as you and @jcvenegas suggested, it would be great to have the ip and port as options.

**Comment by sboeuf** [comment]
@dlespiau , have you got the time to rework your commit according to the comments, or are you planning to do it ?

**Comment by dlespiau** [comment]
Reworking this isn't high priority at the moment, so we either merge it as it and file issues to track what's remaining to do it leaves this pending, so far, it seems we've chosen the second option.

**Comment by sboeuf** [comment]
Let's leave this PR pending until it is reworked then.

**Comment by sameo** [comment]
@jodh-intel @jcvenegas both the HTTP protocol and the 6060 port are defined by the Golang `pprof` package and can not be redefined afaik.
I agree we should warn the user about 2 things:
- This gives internal data through an insecure channel
- We might not be able to bind on 6060

**Comment by jodh-intel** [comment]
@sameo - I wonder why Google haven't had that port registered to make it "more well-known" (well-known outside golang circles)?:

```
$ getent services 6060 || echo not well-known
not well-known
```

Until it is well-known, they should provide an option to change it...

**Comment by sameo** [comment]
LGTM

**Comment by jodh-intel** [comment]
@dlespiau - re https, yes I tend to agree.

lgtm.

**Pull request merged**